### PR TITLE
155 fluka particle source

### PR DESCRIPTION
**Pull request by Bogdan2423**

### CODE REVIEW

**Comment by Bogdan2423** [comment]
Manual testing results on colimator geometry, circular beam located 1.5 cm from colimator, 3 cm radius, energy 70 Mev/nucl
* Proton beam
[fl_sim1.txt](https://github.com/yaptide/converter/files/13400315/fl_sim1.txt)

* Helium-4 beam
[fl_sim2.txt](https://github.com/yaptide/converter/files/13400462/fl_sim2.txt)

File: converter/fluka/cards/beam_card.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        if self.data.shape == BeamShape.GAUSSIAN:
+            shape_what = -1
+            x_y_multiplier = -1
+        elif self.data.shape == BeamShape.CIRCULAR:
+            shape_what = -1
+            x_y_multiplier = 1
+        else:
+            shape_what = 0
+            x_y_multiplier = 1
+        energy = format_float(self.data.energy*-1, 10)
```
Comment (grzanka) [inline converter/fluka/cards/beam_card.py:25]:
```suggestion
        energy = format_float(-self.data.energy, 10)
```

Comment (grzanka) [reply converter/fluka/cards/beam_card.py:25]: Maybe better to rename this quantity to `momentum_or_energy`
The notion of handling something as negative number should very clean to user.
Add here a comment mentioning that `self.data.energy` is storing energy as positive number, explain in which units.
Then say what is stored in `momentum_or_energy`

File: converter/fluka/cards/beam_card.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+        beam_card.what = [energy, 0, 0,
+                          shape_x, shape_y, shape_what]
+        beam_card.sdum = self.data.particle_name
+        if self.data.particle_name == "HEAVYION":
+            hi_card = Card(tag="HI-PROPE")
+            hi_card.what = [self.data.heavy_ion_a, self.data.heavy_ion_z, 0, 0, 0, 0]
+
+        beamposition_card = Card(tag="BEAMPOS")
+        if self.data.z_negative:
```
Comment (grzanka) [inline converter/fluka/cards/beam_card.py:42]: Explain in comment what is in `z_negative`

File: converter/fluka/cards/beam_card.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        pos_x = format_float(self.data.beam_pos[0], 10)
+        pos_y = format_float(self.data.beam_pos[1], 10)
+        pos_z = format_float(self.data.beam_pos[2], 10)
+        dir_x = format_float(self.data.beam_dir[0], 10)
+        dir_y = format_float(self.data.beam_dir[1], 10)
+        beamposition_card.what = [pos_x, pos_y, pos_z,
+                                  dir_x, dir_y, 0]
+        beamposition_card.sdum = z_sdum
+
+        result = f"* {self.data.particle_name} beam of energy {energy*-1} GeV\n"
```
Comment (grzanka) [inline converter/fluka/cards/beam_card.py:56]:
```suggestion
        result = f"* {self.data.particle_name} beam of energy {-energy} GeV\n"
```

In the UI and in most of the tools the energy is in MeV units. I suggest to be clear on the energy unit, i.e. encode unit in the variable name, i.e. using `energy_MeV`. Do the translation to GeV in the final stage of dumping that to text file.

File: converter/fluka/helper_parsers/beam_parser.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    if shape == "Flat circular":
+        return BeamShape.CIRCULAR, shape_params_json["x"], shape_params_json["y"]
+    if shape == "Flat square":
+        return BeamShape.SQUARE, shape_params_json["x"], shape_params_json["y"]
+    if shape == "Gaussian":
+        return BeamShape.GAUSSIAN, shape_params_json["x"], shape_params_json["y"]
+    raise ValueError("Shape type not supported by FLUKA")
+
+
+def cartesian_to_spherical(coords: tuple[float, float, float]):
```
Comment (grzanka) [inline converter/fluka/helper_parsers/beam_parser.py:92]: Isn;t that already in `converter/common.py` ?

File: tests/fluka/test_fluka_beam.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    beam_json["particle"]["id"] = 25
+    beam_json["particle"]["a"] = 2
+    beam_json["particle"]["z"] = 3
+
+    fluka_beam = parse_beam(beam_json)
+
+    assert fluka_beam.energy == 0.14
+    assert fluka_beam.particle_name == "HEAVYION"
+    assert fluka_beam.heavy_ion_a == 2
+    assert fluka_beam.heavy_ion_z == 3
```
Comment (grzanka) [inline tests/fluka/test_fluka_beam.py:35]: Make this check on carbon ion: Z = 6 and A = 12

**Comment by grzanka** [comment]
@Bogdan2423 I've tried with command line and got:

```
ubuntu@yap-dev:~/ui/src/libs/converter$ PYTHONPATH=. python3 converter/main.py editor\(2\).json fl fluka
WARN: number was truncated when converting: 6.123233995736766e-17 -> 0.0
WARN: number was truncated when converting: 6.123233995736766e-17 -> 0.0
File info.json:
{'version': 'unknown', 'label': '', 'simulator': 'fluka'}
File fl_sim.inp:
TITLE
proton beam simulation
* default physics settings for hadron therapy
DEFAULTS                                                              HADROTHE
* HEAVYION beam of energy 0.048 GeV
* flat circular shape with max radius=2.0 cm, min radius=2.0 cm
BEAM          -0.048       0.0       0.0       2.0       2.0      -1.0HEAVYION  
* heavy ion properties: a={self.data.heavy_ion_a}, z={self.data.heavy_ion_z}
HI-PROPE         4.0       2.0       0.0       0.0       0.0       0.0          
* beam position: (0.0, 0.0, 0.0) cm
```

are you sure that comments are printed correctly for heavy ions ?

**Review by grzanka — CHANGES_REQUESTED** [review]
see comments on PR

File: converter/fluka/helper_parsers/beam_parser.py (hunk 4 above)
Comment (Bogdan2423) [reply converter/fluka/helper_parsers/beam_parser.py:92]: at first I wanted to reuse the function from shieldhit converter, so I moved it to common.py, but it turned out Fluka uses spherical coordinates a bit differently, so I wrote this function for Fluka. Apparently I forgot to commit common.py file after moving the function back to shieldhit file

**Review by grzanka — APPROVED** [review]

**Pull request merged**